File: pyrithmetic/Task.py
import math


def normal_pdf(x, mean, std_dev):
    return (1 / (std_dev * math.sqrt(2 * math.pi))) * math.exp(
        -0.5 * ((x - mean) / std_dev) ** 2
    )


def sigmoid(x, shift, scale):
    return 1 / (1 + math.exp(-scale * (x - shift)))


class Task:
    def __init__(self, id, *, accuracy_target, solve_time_target, difficulty_level):
        self.accuracy_target = accuracy_target
        self.solve_time_target = solve_time_target
        self.difficulty_level = difficulty_level
        self.id = id

    def generate_assignment(self):
        raise NotImplementedError("Subclasses must implement generate_assignment()")

    def has_met_targets(self, measured_accuracy, measured_solve_time):
        return (
            measured_accuracy >= self.accuracy_target
            or measured_solve_time <= self.solve_time_target
        )

    def probability_of_selection(self, measured_accuracy, measured_solve_time):
        # Define parameters for normal distribution (mean and standard deviation)
        accuracy_scale = 10  # Adjust as needed
        solve_time_scale = 1  # Adjust as needed

        # difficulty_mean = self.difficulty_level
        # difficulty_std_dev = 1.0  # Adjust as needed

        # Calculate probability using normal distribution PDF
        accuracy_probability = 1 - sigmoid(
            measured_accuracy, shift=self.accuracy_target, scale=accuracy_scale
        )
        solve_time_probability = sigmoid(
            measured_solve_time, shift=self.solve_time_target, scale=solve_time_scale
        )
        # difficulty_probability = normal_pdf(
        #     self.difficulty_level, mean=difficulty_mean, std_dev=difficulty_std_dev
        # )

        # Combine probabilities
        probability = accuracy_probability * solve_time_probability
        return probability

File: js/Task.js
function normalPDF(x, mean, stdDev) {
  return (
    (1 / (stdDev * Math.sqrt(2 * Math.PI))) *
    Math.exp(-0.5 * Math.pow((x - mean) / stdDev, 2))
  );
}

function sigmoid(x, shift, scale) {
  return 1 / (1 + Math.exp(-scale * (x - shift)));
}

export class Task {
  constructor(id, { accuracyTarget, solveTimeTarget }) {
    this.accuracyTarget = accuracyTarget;
    this.solveTimeTarget = solveTimeTarget;
    this.id = id;
  }

  generateAssignment() {
    throw new Error("Subclasses must implement generateAssignment()");
  }

  hasMetTargets(measuredAccuracy, measuredSolveTime) {
    return measuredAccuracy >= this.accuracyTarget;
  }

  probabilityOfSelection(measuredAccuracy, measuredSolveTime) {
    const accuracyScale = 10; // Adjust as needed
    const solveTimeScale = 1; // Adjust as needed

    const accuracyProbability =
      1 - sigmoid(measuredAccuracy, this.accuracyTarget, accuracyScale);
    const solveTimeProbability = sigmoid(
      measuredSolveTime,
      this.solveTimeTarget,
      solveTimeScale,
    );

    const probability = accuracyProbability * solveTimeProbability;
    return probability;
  }
}
